File: java/Solution721.java
package leetCode.java;

import java.util.*;

//Union-Find O(nlogn) O(n), n = 所有種類email的數量
class Solution721 {
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        Map<String, String> parent = new HashMap<>();//<email, parent>
        Map<String, String> user = new HashMap<>();//<email, user>
        for(List<String> account : accounts){
            for(int i=1; i<account.size(); i++){
                parent.put(account.get(i), account.get(i));
                user.put(account.get(i), account.get(0));
            }
        }

        //Union-Find email
        for(List<String> account : accounts){
            String a = find(account.get(1), parent);
            for(int i=2; i<account.size(); i++){
                String b = find(account.get(i), parent);
                parent.put(b, a);
            }
        }

        //過濾掉重複email
        Map<String, Set<String>> merge = new HashMap<>();//<email's parent, <email1, email2, ...>>
        for(List<String> account : accounts){
            for(int i=1; i<account.size(); i++){
                String p = find(account.get(i), parent);
                if(!merge.containsKey(p))
                    merge.put(p, new TreeSet<String>());//emails in sorted order

                merge.get(p).add(account.get(i));
            }
        }

        List<List<String>> res = new ArrayList<>();
        for(Map.Entry<String, Set<String>> entry : merge.entrySet()){
            List<String> line = new ArrayList<>();
            line.add(user.get(entry.getKey()));//放入user
            for(String email : entry.getValue()){
                line.add(email);
            }
            res.add(line);
        }

        return res;
    }

    public String find(String node, Map<String, String> parent){
        if(node.equals(parent.get(node))) return node;
        String p = find(parent.get(node), parent);
        parent.put(node, p);
        return p;
    }
}

File: c++/Solution721.cpp
//Union-Find O(nlogn) O(n), n = 所有種類email的數量
class Solution721 {
public:
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        //一位user可以有多個email, 但一個email只會屬於一位user
        unordered_map<string, string> parents; //<email, email parent>
        unordered_map<string, string> users; //<email, user>
        for (vector<string> account : accounts) {
            string user = account[0];
            for (int i = 1; i < account.size(); ++i) {
                string email = account[i];
                parents[email] = email; // 先設自己為父節點
                users[email] = user; // 另外記錄這個節點email屬於誰
            }
        }

        // union所有email, 這樣有關連的email都會自動到同一組
        for (vector<string> account : accounts) {
            string a = find(account[1], parents);
            for (int i = 2; i < account.size(); ++i) {
                string b = find(account[i], parents);
                parents[b] = a; // 不可以parent[a] = b, 因為a只查一次
            }
        }

        unordered_map<string, set<string>> merged; // <parent, <email, ...>> 透過set 排序與去重複
        for (vector<string> account : accounts) {
            for (int i = 1; i < account.size(); ++i) {
                string parent = find(account[i], parents);
                merged[parent].insert(account[i]);
            }
        }

        vector<vector<string>> res;
        for (auto p : merged) {
            vector<string> account;
            account.push_back(users[p.first]); // username
            account.insert(account.end(), p.second.begin(), p.second.end()); // emails
            res.push_back(account);
        }
        return res;
    }

    string find(string node, unordered_map<string, string>& parents) {
        if (parents[node] == node) return node;
        return parents[node] = find(parents[node], parents);
    }
};